### PR TITLE
Refactored Perceptionhash for performance

### PR DESCRIPTION
**Pull request by evanoberholster**
Refactored PerceptionHash for performance. If you want me to change the name to replace existing function that would be acceptable.

Features:
- Significantly reduces allocations
- Uses static DCT1D tables
- Uses sync pool to reduce pixel data allocations
- Uses image.Image optimizations for obtaining pixels for improved inlining of functions
- Backward compatible with PerceptionHash, they are interchangeable. Confirmed with passing tests.
- Performance without image resize as below.
```go
name                      time/op
PerceptionHash/Regular-12         358µs ±26%
PerceptionHash/Fast-12           84.8µs ± 7%
PerceptionHash/Parallel-12       82.5µs ± 3%
PerceptionHash/Fast-Parallel-12  11.5µs ± 1%

name                      alloc/op
PerceptionHash/Regular-12         193kB ± 0%
PerceptionHash/Fast-12            17.0B ± 0%
PerceptionHash/Parallel-12        193kB ± 0%
PerceptionHash/Fast-Parallel-12    198B ± 0%

name                      allocs/op
PerceptionHash/Regular-12         4.68k ± 0%
PerceptionHash/Fast-12             0.00     
PerceptionHash/Parallel-12        4.68k ± 0%
PerceptionHash/Fast-Parallel-12    0.00 
```

### CODE REVIEW

**Comment by evanoberholster** [comment]
Comparison between PerceptionHash and PerceptionHashFast:
```go
name               old time/op    new time/op    delta
PerceptionHash-12     945µs ± 9%     689µs ± 2%  -27.08%  (p=0.000 n=170+177)

name               old alloc/op   new alloc/op   delta
PerceptionHash-12     471kB ± 0%     298kB ± 0%  -36.80%  (p=0.000 n=174+198)

name               old allocs/op  new allocs/op  delta
PerceptionHash-12     4.74k ± 0%     0.07k ± 0%  -98.61%  (p=0.000 n=200+200)
```

**Review by corona10 — CHANGES_REQUESTED** [review]
@evanoberholster Looks nice! Thank you for your contribution!!

>  If you want me to change the name to replace existing function that would be acceptable.

I prefer this one(change the name ) :)

**Comment by corona10** [comment]
Please add your name to the following file.
https://github.com/corona10/goimagehash/blob/master/AUTHORS.md

**Comment by corona10** [comment]
> Significantly reduces allocations
> Uses static DCT1D tables
> Uses sync pool to reduce pixel data allocations
> Uses image.Image optimizations for obtaining pixels for improved inlining of functions
> Backward compatible with PerceptionHash, they are interchangeable. Confirmed with passing tests.
> Performance without image resize as below.

Thanks, I love all of your changes :)

File: hashcompute.go [hunk 1]
```diff
@@ -84,6 +85,44 @@ func PerceptionHash(img image.Image) (*ImageHash, error) {
...
+		p := make([]float64, 4096)
+		return &p
+	},
+}
+
+// PerceptionHashFast function returns a hash computation of phash.
+// Uses static DCT tables for improved performance.
+// Implementation follows
+// http://www.hackerfactor.com/blog/index.php?/archives/432-Looks-Like-It.html
+func PerceptionHashFast(img image.Image) (*ImageHash, error) {
```
Comment (corona10) [inline hashcompute.go:99]:
```suggestion
func PerceptionHash(img image.Image) (*ImageHash, error) {
```

**Review by corona10 — CHANGES_REQUESTED** [review]
Would you like to switch the PerceptionHashFast into PerceptionHash and remove old implementation?

**Comment by evanoberholster** [comment]
PerceptionHashFast DCT static tables only support a size of 64x64. We would need to include further static tables to support other image sizes, (eg 256x256)

**Comment by corona10** [comment]
@evanoberholster 

> PerceptionHashFast DCT static tables only support a size of 64x64.

Ah okay, let's replace PerceptionHash only.

**Comment by evanoberholster** [comment]
Sorry for the delay I have been busy at work. Will get it replaced and update the PR.

**Comment by corona10** [comment]
> Sorry for the delay I have been busy at work. Will get it replaced and update the PR.

Yeah, Don't worry I can wait :)

File: hashcompute.go (hunk 1 above)
Comment (evanoberholster) [reply hashcompute.go]: Made these changes in new update.

File: hashcompute_test.go [hunk 2]
```diff
@@ -98,6 +102,78 @@ func TestHashCompute(t *testing.T) {
...
+
+		if dis1 != dis2 {
+			t.Errorf("Distance should be identical %v vs %v", dis1, dis2)
+		}
+		if dis2 != 0 {
+			t.Errorf("Distance should be identical %v vs %v, with distance of %d", hash1.ToString(), hash2.ToString(), dis1)
+		}
+	}
+}
+
```
Comment (corona10) [inline hashcompute_test.go:176]: IMHO, we can remove this test since `TestHashCompute` already covers this area.

```suggestion
```

**Review by corona10 — CHANGES_REQUESTED** [review]
Almost looks good to me, Just 1 nit comment and I will take a look at more later :)
Thanks for the change!

**Comment by evanoberholster** [comment]
I agree that those tests are redundant. Thanks.

**Review by corona10 — APPROVED** [review]

File: hashcompute_test.go [hunk 3]
```diff
@@ -9,6 +9,10 @@ import (
 	"image/jpeg"
 	"os"
 	"testing"
+
+	"github.com/corona10/goimagehash/etcs"
+	"github.com/corona10/goimagehash/transforms"
+	"github.com/nfnt/resize"
```
Comment (corona10) [inline hashcompute_test.go:15]:
```suggestion

```

Should be removed :)

**Pull request merged**